### PR TITLE
Improve Creature DBC Handling

### PR DESCRIPTION
**Pull request by HelloKitty**
**Changes proposed:**

-  Remove hardcoded literal flags check for CreatureModelData mount capability.
-  Check for CreatureTypeEntry flags for calculating Creature::CanGiveExperience and remove IsCritter check (defined in DBC along with gascloud and non-combat pets)

**Target branch(es):** 3.3.5/master

- [x] 3.3.5
- [ ] master

**Issues addressed:** Closes #  (insert issue tracker number)

**Tests performed:** (Does it build, tested in-game, etc.)

Can still gain experience from normal creatures. Cannot get experience from critters.

**Known issues and TODO list:** (add/remove lines as needed)

None

### CODE REVIEW

**Comment by HelloKitty** [comment]
The build failure is not from my commit, it's from: ee5eace

**Comment by funjoker** [comment]
Please revert to your commit and use rebase

**Comment by HelloKitty** [comment]
I tried running a rebase but I must have done it wrong. I'm unfamiliar with this area of git. If I keep trying to fix it then it'll probably get worse, like the other PR I opened. I'm sorry.

**Comment by HelloKitty** [comment]
Ok, small chance I may have fixed it.

File: src/server/game/Entities/Creature/Creature.cpp [hunk 1]
```diff
@@ -3245,8 +3245,12 @@ void Creature::ClearTextRepeatGroup(uint8 textGroup)
 
 bool Creature::CanGiveExperience() const
 {
-    return !IsCritter()
-        && !IsPet()
+    CreatureTypeEntry const* creatureType = sCreatureTypeStore.AssertEntry(GetCreatureType());
+
+    if (creatureType->HasFlag(CreatureTypeEntryFlags::CREATURE_TYPE_ENTRY_FLAGS_NO_EXPERIENCE))
```
Comment (Shauren) [inline src/server/game/Entities/Creature/Creature.cpp:3250]: You should revert this change, this flag has a different meaning, its a "do not include in tab targeting"

Comment (HelloKitty) [reply src/server/game/Entities/Creature/Creature.cpp:3250]: Ohhh ok, that actually makes a lot of sense now. The information I was basing it on claimed it was something else.

Comment (HelloKitty) [reply src/server/game/Entities/Creature/Creature.cpp]: Partially reverted that commit now.

File: src/server/shared/DataStores/DBCStructure.h [hunk 2]
```diff
@@ -522,12 +529,19 @@ struct CreatureSpellDataEntry
...
+    CREATURE_TYPE_ENTRY_FLAGS_IGNORED_TAB_TARGETING = 0x01 // Means do not include in tab targeting.
+};
+
 struct CreatureTypeEntry
 {
     uint32    ID;                                           // 0        m_ID
     //char*   Name[16];                                     // 1-16     name
                                                             // 17       string flags
-    //uint32    no_expirience;                              // 18 no exp? critters, non-combat pets, gas cloud.
+    uint32    Flags;                                        // 18 Critters, non-combat pets, gas cloud. (called Flags in struct)
```
Comment (Shauren) [inline src/server/shared/DataStores/DBCStructure.h:542]: There is no reason to load this field anymore

File: src/server/game/Entities/Unit/Unit.cpp [hunk 3]
```diff
@@ -8902,7 +8902,7 @@ bool Unit::IsInDisallowedMountForm() const
     CreatureModelDataEntry const* model = sCreatureModelDataStore.LookupEntry(display->ModelId);
     ChrRacesEntry const* race = sChrRacesStore.LookupEntry(displayExtra->Race);
 
-    if (model && !(model->Flags & 0x80))
+    if (model && !(model->HasFlag(CreatureModelDataFlags::CREATURE_MODEL_DATA_FLAGS_CAN_MOUNT)))
```
Comment (Shauren) [inline src/server/game/Entities/Unit/Unit.cpp:8905]: Codestyle remark: dont prefix regular enums with `Type::`, use that only for `enum class` where its required

File: src/server/shared/DataStores/DBCStructure.h (hunk 2 above)
Comment (HelloKitty) [reply src/server/shared/DataStores/DBCStructure.h:542]: True, but why is so much effort put into not loading parts of DBC? Their memory footprint seems to be nearly nothing.

Comment (Shauren) [reply src/server/shared/DataStores/DBCStructure.h:542]: I have no idea

**Comment by jackpoz** [comment]
could you merge 3.3.5 branch into this PR ?

**Comment by HelloKitty** [comment]
@jackpoz Last time I tried to do that I messed it up and lost some of my sanity.

**Comment by jackpoz** [comment]
didn't you do a rebase last time ? I'm asking for a merge, that should be easier
- checkout 335
- pull TC 335
- switch to the PR branch
- merge 335 into the PR branch

**Comment by jackpoz** [comment]
Thanks for the PR :)
